### PR TITLE
Add constructs for safe handling of PII.

### PR DESCRIPTION
**Pull request by mcsaucy**
This change centralizes PII handling with a series of wrapper classes.
These classes selectively store or redact data based upon the
`--collect_pii` flag. This lets the actual scraping code just grab
everything, throw it in the PII wrappers and trust that they'll Do The
Right Thing, without materially changing the usage of the PII wrapped
data. This is accomplished by subclassing the data we wish to
encapsulate. A `Pii.String` is a subclass of `str`, for example.

PII handling subclasses are cool and all, but they aren't really useful
if they don't get consistently used. To that end, this change alters our
`Record` dataclass to include complete type hints for all fields and
ensures the field values match the type at runtime. Attempting to shove
a normal string into a `Pii.String`-typed field will result in a
`TypeError`, for example.

While we're here, alter our `Record` and `Charge` definitions to call
out optional fields and expose builder objects for them. Given the
number of fields these have, this is less error prone than setting a
bunch of individual variables and plugging them into our dataclass
constructors at the end.

### CODE REVIEW

**Comment by salter31** [comment]
@mcsaucy I haven't done a deep review of the entire repo but from a code maintenance perspective it would be safer and easier to maintain if the PII checks were done at the [shared record](https://github.com/Police-Data-Accessibility-Project/Scrapers/blob/c4393eb06bc4b648aa85e3f913c0254e1ad7b762/USA/FL/Bay/Court/scraper/utils/ScraperUtils.py#L29) level instead of the scraper level. I don't think that class is shared across scrapers but ideally there would be 1 layer that handles anonymization so we don't have to integrate the PII class across every single integration that gets completed. Thats how leaks and security holes are created.

**Comment by mcsaucy** [comment]
We don't currently have "1 layer" to do this. Would moving our record definition into `common` so other scrapers can use it satisfy that requirement? Are you cool with this approach otherwise?

**Comment by OscarVanL** [comment]
The only issue I can foresee is that the fields collected might vary by scraper. 

To give one example, I felt for the Benchmark portal it was helpful to store both the Florida "Uniform Case Numbering System" Case ID as well as the Portal's Case ID, and there was an "Agency Report Number" field on the website, which I thought could be helpful for cross-referencing cases to police report numbers in case we ever get our hands on police report dataset. 

These were not originally listed in the [minimum scraped fields list](https://github.com/Police-Data-Accessibility-Project/Police-Data-Accessibility-Project#the-fields-we-would-like-to-make-sure-to-collect-at-a-minimum-from-any-scrape-are).

Basically, we cannot necessarily be sure `Record` and `Charge` will have the same fields for each state. 

In principle a shared `common` definition has advantages, but the ability to extend the dataclasses with extra fields may be necessary.

**Comment by mcsaucy** [comment]
So do we think there's value in a dataclass with the minimal fields (and PII types) in `common` that scrapers can extend as needed?

**Comment by OscarVanL** [comment]
Yes, that sounds like a good idea. 

In practice can you extend dataclasses?

So could I go:
```
@dataclass
class BenchmarkRecord(Record):
    new_field: type
```

to add new fields ontop of those minimal fields/PII types?

File: USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py [hunk 1]
```diff
@@ -57,12 +59,22 @@ def test_parse_plea_type_not_guilty(self):
...
         assert ScraperUtils.parse_name(name2) == ('JOHN', None, 'DOE')
 
+    @flagsaver.flagsaver(collect_pii=False)
+    def test_parse_name_redacted(self):
+        name1 = "DOE, JANE EMILY"
+        name2 = "DOE, JOHN"
+        assert ScraperUtils.parse_name(name1) == (
+                '[redacted]', '[redacted]', '[redacted]')
+        assert ScraperUtils.parse_name(name2) == (
+                '[redacted]', '[redacted]', '[redacted]')
```
Comment (OscarVanL) [inline USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py:76]: I like this. I agree that it's more transparent to write '[redacted]' than to leave the fields blank.

File: USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py [hunk 2]
```diff
@@ -89,19 +101,20 @@ def test_parse_charge_statute_incomplete(self):
         assert ScraperUtils.parse_charge_statute(' ') == (None, None)
         assert ScraperUtils.parse_charge_statute(None) == (None, None)
 
+    @flagsaver.flagsaver(collect_pii=True)
     def test_parse_defense_attorneys(self):
         attorneys1 = ['DEFENSE ATTORNEY: DOE, JANE EMILY ASSIGNED', 'DEFENSE ATTORNEY: DOE, JOHN MICHAEL ASSIGNED', 'DEFENSE ATTORNEY: SELF, SELF ASSIGNED']
         public_defenders1 = ['COURT APPOINTED ATTORNEY: DOE, JOHN MICHAEL ASSIGNED']
-        assert ScraperUtils.parse_attorneys(attorneys1) == ['DOE, JANE EMILY', 'DOE, JOHN MICHAEL', 'SELF, SELF']
-        assert ScraperUtils.parse_attorneys(public_defenders1) == ['DOE, JOHN MICHAEL']
+        assert ScraperUtils.parse_attorneys(attorneys1) == ('DOE, JANE EMILY', 'DOE, JOHN MICHAEL', 'SELF, SELF')
```
Comment (OscarVanL) [inline USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py:108]: Is this correct? parse_defense_attorneys is supposed to return a list of strings [], not a tuple ().

File: USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py [hunk 3]
```diff
@@ -139,3 +152,24 @@ def test_parse_out_path_correct_length(self):
...
+                                first_name="I should be in a PII wrapper")
+
+    @flagsaver.flagsaver(collect_pii=False)
+    def test_happy_builder(self):
+        builder = ScraperUtils.RecordBuilder(
+                id="foo", state="bar", county="baz")
+        want = Pii.String("I'm in a PII wrapper")
+        builder.first_name = want
+        r = builder.build()
+        assert r.first_name == want
```
Comment (OscarVanL) [inline USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py:175]: Add unit tests for the expected behaviour if collect_pii=True?

File: USA/FL/Bay/Court/scraper/utils/ScraperUtils.py [hunk 4]
```diff
@@ -1,57 +1,68 @@
...
-    case_status: str
-    defense_attorney: str
-    public_defender: str
-    judge: str
-    charges: List[Charge]
-    arresting_officer: str
-    arresting_officer_badge_number: str
+    id: Optional[str]
+    state: Optional[str]
+    county: Optional[str]
```
Comment (OscarVanL) [inline USA/FL/Bay/Court/scraper/utils/ScraperUtils.py:40]: Any practical reasons these fields are Optional, since they need to be used for DB keys?
Surely these should be the minimum required fields?

File: common/pii/Pii.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+from typing import Optional, Generic, TypeVar, Sequence, Tuple
+
+from absl import flags
+
+FLAGS = flags.FLAGS
+flags.DEFINE_bool("collect_pii", False, "Whether to store PII.")
```
Comment (OscarVanL) [inline common/pii/Pii.py:6]: I assume this won't override the user's own `collect_pii` setting set in arguments as `common.pii.Pii` is imported before args are parsed. 

Is this is just to ensure the default setting is False?

File: USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py (hunk 2 above)
Comment (mcsaucy) [reply USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py:108]: I went with tuple due to immutability. Otherwise one could make a Pii.StringSequence and then add raw PII to it without going through the --collect_pii check.

File: common/pii/Pii.py (hunk 5 above)
Comment (mcsaucy) [reply common/pii/Pii.py:6]: Yeah, `False` is just the default.

File: USA/FL/Bay/Court/scraper/utils/ScraperUtils.py (hunk 4 above)
Comment (mcsaucy) [reply USA/FL/Bay/Court/scraper/utils/ScraperUtils.py:40]: Nope, this was a mistake. Made them mandatory again.

File: USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py (hunk 3 above)
Comment (mcsaucy) [reply USA/FL/Bay/Court/scraper/tests/test_ScraperUtils.py]: Done.

**Comment by mcsaucy** [comment]
> In practice can you extend dataclasses?

Looks like it. I've made a base record in `common` and extended it with the extra fields we're tracking.

**Comment by OscarVanL** [comment]
LGTM! 

Looks like it does everything we want it to do, it provides a shared record as suggested by @salter3, it necessitates all required DB keys, it allows for the protection of PII fields and it allows for extensibility with extra portal-specific fields.

Let's squash and merge :)

**Comment by mcsaucy** [comment]
Turns out Github's "squash and merge" option works well.
